Authentication

rusher opened the pull request:
### Description
authentication has been revamp. this is a simplification of https://github.com/go-sql-driver/mysql/pull/1694, without new authentication plugin addition, since it's already a big change.

Current implementation was requiring authentication switch plugins data to begin with iAuthMoreData, but that's only required for "caching_sha2_password" and "sha256_password" plugins.
Additionally, now permits multi-authentication (in mariadb: https://mariadb.com/kb/en/create-user/#identified-viawith-authentication_plugin / mysql: https://dev.mysql.com/doc/refman/8.0/en/multifactor-authentication.html)

goal is to add MariaDB main authentication plugins (like parsec, PAM, GSSAPI, ...) based on plugin like this

### Checklist
- [x] Code compiles correctly
- [x] Created tests which fail without the change (if possible)
- [x] All tests passing
- [x] Extended the README / documentation, if necessary
- [x] Added myself / the copyright holder to the AUTHORS file

methane commented:
Make "test stability improvement."  PR first.

methane commented on auth_plugin.go:
```diff
@@ -0,0 +1,63 @@
+
+// AuthPlugin represents an authentication plugin for MySQL/MariaDB
+type AuthPlugin interface {
+	// GetPluginName returns the name of the authentication plugin
+	GetPluginName() string
```
```suggestion
	// PluginName returns the name of the authentication plugin
	PluginName() string
```

rusher replied:
corrected

methane commented on auth_plugin.go:
```diff
@@ -0,0 +1,63 @@
+	// packet is the data from the server's auth response
+	// authData is the initial auth data from the server
+	// conn is the MySQL connection (for performing additional interactions if needed)
+	ContinuationAuth(packet []byte, authData []byte, conn *mysqlConn) ([]byte, error)
```
Can external plugins use `mysqlConn`?

We should not expose plugin interface until its design is complete and stable.

rusher replied:
changed

methane commented:
Maybe, its time to use internal package.

rusher commented on auth.go:
```diff
@@ -35,27 +28,6 @@ var (
-//		mysql.RegisterServerPubKey("mykey", rsaPubKey)
-//	} else {
-//		log.Fatal("not a RSA public key")
-//	}
```
yes, corrected

rusher commented on auth_plugin.go:
```diff
@@ -0,0 +1,63 @@
+}
+
+// PluginRegistry is a registry of available authentication plugins
+type PluginRegistry struct {
```
corrected

rusher commented on auth_plugin.go:
```diff
@@ -0,0 +1,63 @@
+}
+
+// NewPluginRegistry creates a new plugin registry
+func NewPluginRegistry() *PluginRegistry {
```
corrected

rusher commented on driver.go:
```diff
@@ -39,8 +39,9 @@ type DialFunc func(addr string) (net.Conn, error)
-	dials     map[string]DialContextFunc
+	dialsLock            sync.RWMutex
+	dials                map[string]DialContextFunc
+	globalPluginRegistry = NewPluginRegistry()
```
cannot do since that's the name of the struct

rusher commented:
rebase on current master, and corrected the point above

rusher commented on auth_caching_sha2.go:
```diff
@@ -0,0 +1,187 @@
+	case iAuthMoreData:
+		switch len(packet) {
+		case 1:
+			return mc.readPacket() // Auth successful
```
i've keeped this part because was existing, but this is not in protocol https://dev.mysql.com/doc/dev/mysql-server/latest/page_caching_sha2_authentication_exchanges.html, server is expected to either send result fast_auth_success(3) or perform_full_authentication(4) only

rusher commented:
Do you have any thoughts or concerns about when this can be merged?

methane commented:
since this PR affects public API, I need a lot of time for DeepThink.
No schedule for now.

grooverdan commented:
@methane, how can I help?

methane commented:
This PR uses mysqlConn in the Public Interface. So I need to be able to answer these questions myself.
"Is this design correct? Can other users use this interface? Can this project support this API forever?"

If you want to move this PR forward, please limit it to internal structural changes that do not add any public API, or define an Interface that provides all the functionality required by the auth plugin instead of mysqlConn and use that.

methane commented on auth_plugin.go:
```diff
@@ -0,0 +1,63 @@
+	// packet is the data from the server's auth response
+	// authData is the initial auth data from the server
+	// conn is the MySQL connection (for performing additional interactions if needed)
+	continuationAuth(packet []byte, authData []byte, conn *mysqlConn) ([]byte, error)
```
@copilot このinterfaceはpublicですが、 continuationAuth は privateで、その引数のmysqlConn型もprivateです。このinterfaceを外部から利用するのはどう言う状況ですか？

rusher committed:
Authentication has been revamp, permitting authentication plugin. #1696

now permits multi-authentication (in mariadb: https://mariadb.com/kb/en/create-user/#identified-viawith-authentication_plugin / mysql: https://dev.mysql.com/doc/refman/8.0/en/multifactor-authentication.html)

aldebaraner commented:
Force-pushed corrections based on master.

I agree that exposing mysqlConn in the API is not ideal. I've reworked the authentication plugin interface to address this concern.

New signature:
```
continuationAuth(packet []byte, authData []byte, cfg *Config) (nextPacket []byte, done bool, err error)
```

Changes from the previous version:

The driver now manages all protocol communication. Plugins operate purely on packet data, and return packet data to be sent. 
Driver reads incoming packets and routing them appropriately, either dispatching to the plugin for continuation or processing them through the normal protocol flow
This keeps the plugin interface clean and focused on authentication logic only, without requiring direct access to connection internals.

rusher commented:
@methane I think this is ok, have you any opinion ? can i add directly some mariadb authentiction plugins ? (parsec, dialog, ...)

mariadb-AllenHerrera commented:
Hey @methane ,

I am a Product Manager at MariaDB. First, thank you for your incredible work maintaining this driver, it is a critical piece of infrastructure in the GO ecosystem.

We are looking for ways to support the project and work closer with you to help the driver evolve alongside MariaDB's latest features. We'd love to discuss how we can help you with maintenance, faster iteration on PRs, or providing testing resources to benefit the community.

Would you be open to a brief meeting with our engineering lead, @rusher, and me to discuss how we can best support you?

rusher committed:
Add validation and thread-safety improvements to authentication plugin system

rusher committed:
Improve error handling and refactor authentication plugin system

methane commented:
@mariadb-AllenHerrera 
The most important thing to make this project sustainable is to avoid adding features that are not truly needed by a large number of people.

I know that MariaDB adds many useful features to the MySQL client/server protocol. However, many managed databases (Azure, Google Cloud SQL, AWS Aurora) ignore MariaDB and only support MySQL.
Therefore, it is necessary to carefully consider whether or not to support features that are only available in MariaDB.

About #1689, we will consider reviewing this feature if you submit a pull request. The overhead of three round trips is significant.

For other features, could you please suggest them on the forum?
https://github.com/go-sql-driver/mysql/discussions/categories/ideas

The best thing would be to communicate with the MySQL developers and get them to incorporate MariaDB extensions into MySQL. If both #1708 and #1689 were supported in MySQL, it would be a huge benefit for all users.

mariadb-AllenHerrera commented:
Hi @methane ,

Thank you for the response and for being open to reviewing the PRs. We fully agree that performance and simplicity are vital to a project's sustainability.

Our goal isn't to break the protocol, but to make the driver environment-aware. It should remain a lean MySQL client by default, while seamlessly unlocking advanced capabilities when it detects MariaDB.

We understand the importance of prioritizing the majority of users, and the managed services you mentioned carry real weight. But they don't define the full picture. MariaDB is the default MySQL-compatible database on every major Linux distribution: Debian, Ubuntu, RHEL, Fedora, SUSE, and has surpassed MySQL for WordPress backends. Our docker downloads surpassed 3 billion downloads, and many self-hosted Go workloads likely already run against MariaDB but not at its full capability via your project and others.

**Why this is a win for the project**:

**User Retention**: With many users migrating to MariaDB, adaptability prevents them from having to choose between their trusted driver and the features they need. Letting us help ensures the project remains the universal standard for the Go ecosystem.

**Performance Neutrality**: We are committed to zero overhead for MySQL users; the "MySQL path" will remain as fast as it is today ideally.

**Future Proofing**: Robust handshake and capability-sensing make the driver resilient to changes in any future fork or version.

**Regarding the diverged protocols**: Oracle has no commercial incentive to adopt MariaDB features. Thus, we must support the millions of users in the MariaDB production ecosystem now, rather than relying on an upstream path that is effectively closed.

We'd love to work with you to ensure these changes meet your standards. While we are prepared to fork to ensure the MariaDB community isn't left behind, we believe a single, versatile driver is the best outcome for the entire community.

Best regards,
Allen Herrera

rusher committed:
Implement secure transport requirement for cleartext authentication plugins

rusher committed:
Merge remote-tracking branch 'origin/master' into authentication

rusher commented:
added TLS/unix socket security fix to permit sending clear text password + merge with current head

rusher committed:
Add unix socket configuration for mysql_clear_password in multi-auth tests

rusher commented:
_remaining failing tests are already present on master branch_

methane commented on auth.go:
```diff
@@ -84,401 +77,140 @@ func getServerPubKey(name string) (pubKey *rsa.PublicKey) {
+// to the appropriate auth plugin.
+func (mc *mysqlConn) handleAuthResult(remainingSwitch uint, initialSeed []byte, authPlugin AuthPlugin) error {
+	if remainingSwitch == 0 {
+		return fmt.Errorf("maximum of %d authentication switch reached", authMaximumSwitch)
```
Move the check to the auth switch packet processing.

rusher replied:
right, corrected

rusher committed:
Move auth switch limit check to switch packet processing

rusher commented:
any news ?
